Feature/exclude folders from cleanup

a-custodio opened the pull request:
This PR enhances the project cleanup module with two key safety and control features:

Folder Exclusion: Adds a `target_excluded_folders` variable to prevent the function from processing specified folders and their contents. This is crucial for protecting sensitive environments from accidental cleanup.

Dry Run Mode: Introduces a `dry_run` mode, enabled by default, which logs all intended deletions without performing them. This allows for safe validation of the function's behavior before enabling destructive operations.

a-custodio committed:
feat(cleanup): Add support for excluding folders from cleanup

This commit introduces a new variable `target_excluded_folders` to the
project_cleanup module.

This allows users to provide a list of folder IDs that should be
completely ignored by the cleanup function. The variable is passed as a
JSON-encoded environment variable `TARGET_EXCLUDED_FOLDERS` to the
Cloud Function.

a-custodio committed:
feat(function): Implement folder exclusion logic in cleanup function

The cleanup function now reads the `TARGET_EXCLUDED_FOLDERS` environment
variable and parses it into a map for efficient lookups.

The recursive folder traversal function,
`getSubFoldersAndRemoveProjectsFoldersRecursively`, has been updated to
check if a folder is in the exclusion list before processing it. If a
folder is marked for exclusion, the function will not recurse into its
sub-folders or process any of its projects, effectively skipping that
entire branch of the resource tree.

a-custodio committed:
feat(project_cleanup): add dry_run mode to cleanup function

Adds a 'dry_run' mode to the project cleanup function. When enabled, the function will log the actions it would take (e.g., deleting projects or folders) without actually performing any destructive operations.

This provides a safer way to test and verify the function's behavior before applying actual changes to the GCP organization.

- A `dry_run` boolean variable was added to the Terraform module, defaulting to `true`.
- This variable is passed as a `DRY_RUN` environment variable to the Go Cloud Function.
- The Go function now checks for this variable and skips the deletion logic if it's enabled, logging the intended actions instead.

a-custodio committed:
fix(project_cleanup): correct the variable name for dry_run mode

Resolves a compilation error in the Go Cloud Function caused by an incorrect variable name.

The environment variable was being read into a global variable named `isDryRun`, but the function logic was attempting to reference a non-existent variable `dryRun`.

a-custodio committed:
docs(project_cleanup): update READMEs for new features

Manually updates the documentation to include the recently added `dry_run` and `target_excluded_folders` functionalities.

- `modules/project_cleanup/README.md`: The "Inputs" table is updated with the new Terraform variables.
- `modules/project_cleanup/function_source/README.md`: The "Environment Configuration" table is updated to reflect the new environment variables used by the Go function.

a-custodio committed:
docs(project_cleanup): reorder README inputs alphabetically

Corrects the previous documentation update by sorting the "Inputs" table in the module's README.md alphabetically.

guilherme-hosoda-cit commented on modules/project_cleanup/variables.tf:
```diff
@@ -154,3 +154,15 @@ variable "function_docker_registry" {
+}
+variable "target_excluded_folders" {
+  description = "List of folder IDs (e.g., '1234567890') to exclude from the entire cleanup process."
+  type        = list(string)
```
Change the variable type to `set(string)` instead of `list(string)` to skip duplicated folders.

guilherme-hosoda-cit reviewed (approved):
LGTM

a-custodio committed:
refactor(project_cleanup): Use set for excluded folders variable

Changes the type of the `target_excluded_folders` variable from `list(string)` to `set(string)`.

This makes the module more robust by automatically handling any duplicate folder IDs provided by the user, ensuring the input is always a unique collection of folders.
